sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1050 @@
+
+2. **Queue recording API (aka "Record & Replay")** - Introduces state to a
+`sycl::queue` such that rather than scheduling commands immediately for
+execution, they are added to the graph object instead, with edges based on the
+data dependencies of the command group.
+
+Each of these mechanisms for constructing a graph have their own advantages, so
+having both APIs available allows the user to pick the one which is most
+suitable for them. The queue recording API allows quicker porting of existing
+applications, and can capture external work that is submitted to a queue, for
+example via library function calls. While the explicit API can better express
+what data is internal to the graph for optimization, and dependencies don't need
+to be inferred.
+
+It is valid to combine these two mechanisms sequentially when constructing a
+graph, however it is not valid to use them concurrently. An error will be thrown
+if a user attempts to use the explicit API to add a node to a graph which is
+being recorded to by a queue.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_GRAPH` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+=== SYCL Graph Terminology
+
+:explicit-memory-ops: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:explicitmemory
+
+Table 2. Terminology.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Graph
+| A directed and acyclic graph (DAG) of commands (nodes) and their dependencies
+(edges), represented by the `command_graph` class.
+
+| Node
+| A command, which can have different attributes.
+
+| Edge
+| Dependency between commands as a happens-before relationship.
+
+|===
+
+==== Explicit Graph Building API
+
+When using the explicit graph building API to construct a graph, nodes and
+edges are captured as follows.
+
+Table 3. Explicit Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| In the explicit graph building API nodes are created by the user invoking
+methods on a modifiable graph. Each node represent either a command-group
+function, empty operation, or device memory allocation/free.
+
+| Edge
+| In the explicit graph building API edges are defined by the user. This is
+either through buffer accessors, the `make_edge()` function, or by passing
+dependent nodes on creation of a new node.
```

Comment:
Do event dependencies (i.e., `handler::depends_on()`) also create an edge?

Reply:
For now, we won't allow dependencies upon any events for commands within a graph. It's a feature we would consider for future development of the proposal. I'll add those restrictions. What you can do is either wait on an event before the graph is executed. Or wait upon completion of a graph by using the returned event of `submit`.

Reply:
How does an application that uses USM pointers (instead of buffers and accessors) then express dependencies between the different kernels in the graph? 

For the explicit graph building API, applications can use `make_edge()`, but what about the graph recording API? How can applications specify dependencies between kernels using USM pointers with the recording API?

Reply:
@bensuo had thoughts about `handler::depends_on()` so he could possibly explain better, but I think it is the ability of the function to capture arbitrary events which aren't associated to graph nodes which turned us off this functionality.

USM dependencies are indeed more awkward with the record & replay API. Where we got to in the ComputeCpp implementation of them, was that USM pointers in kernel arguments had to be the same for the runtime to create an edge between the kernels. And I think if the runtime did more work it could connect offset USM pointers used as parameters to the same allocation, make that an edge. However, this isn't perfect, as for system USM allocations (`malloc`/`new`) the SYCL runtime won't know the size of an allocation, so I don't think could work out offsets to the same allocation. We've not been too descriptive in the spec language about the USM dependencies tracking for record & replay, and that's been because it feels like it's getting into implementation details.

Tracking `handler::depends_on()` for record & replay would remove the system USM limitation, and mean the runtime has to do less work to bookkeep and search active USM allocations. We currently do return a `sycl::event` from recorded queue commands, but it is specified as a default constructed event. So I think to achieve using `handler::depends_on()`  we'd
need to a have a way to have a `sycl::event` correspond to a graph node, and then we can specify that for `handler::depends_on()` only these special events are used as edges.

Reply:
> So the issue arises if nodes of the graph depend on events of commands that are not part of the graph?

Yes, I think my concern is mostly conceptual. You define a `command_graph` and then execute it, but now we're introducing nodes into the graph which don't have this model. So you won't be able to represent the non-`command_graph` commands in the Level Zero command-list (to pick a backend API) of the graph. And you'd probably need to split your graph into multiple command-lists at points where there's an event to a non-`command_graph` command. This partitioning will be needed in other cases too, so it's not impossible.

I actually have a PR to change `queue::submit(command_graph<graph_state::executable> graph);`  -
https://github.com/reble/llvm/pull/26

Reply:
> So the issue arises if nodes of the graph depend on events of commands that are not part of the graph?

Yes I think this is the primary issue. Conceptually I think if a specific node depends on an event outside of the graph then whatever is creating that event is part of the graph and should be recorded as such. Alternatively as @EwanC mentions it could be added as a dependency to the graph submission when actually submitting the graph if it is some one time event that needs to happen before the graph actually executes